File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lyricsscraper/activity/SettingsActivity.java
package com.wa2c.android.medoly.plugin.action.lyricsscraper.activity;

import android.app.ActionBar;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.preference.EditTextPreference;
import android.preference.ListPreference;
import android.preference.MultiSelectListPreference;
import android.preference.Preference;
import android.preference.PreferenceActivity;
import android.preference.PreferenceCategory;
import android.preference.PreferenceFragment;
import android.preference.PreferenceScreen;
import android.provider.Settings;
import android.text.InputType;
import android.text.TextUtils;
import android.view.MenuItem;

import com.wa2c.android.medoly.plugin.action.lyricsscraper.R;
import com.wa2c.android.medoly.plugin.action.lyricsscraper.dialog.AboutDialogFragment;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Set;


/**
 * Settings activity.
 */
public class SettingsActivity extends PreferenceActivity {

    /**
     * onCreate event.
     */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getFragmentManager().beginTransaction().replace(android.R.id.content, new SettingsFragment()).commit();

        // Action bar
        ActionBar actionBar = getActionBar();
        if (actionBar != null) {
            actionBar.setDisplayShowHomeEnabled(true);
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowTitleEnabled(true);
            actionBar.setTitle(R.string.title_activity_settings);
        }
    }

    /**
     * onOptionsItemSelected event.
     */
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
        }

        return super.onOptionsItemSelected(item);
    }



    /**
     * Settings fragment.
     */
    public static class SettingsFragment extends PreferenceFragment {

        /** Summary length. */
        private static final HashMap<Preference, Integer> summaryLengthMap = new LinkedHashMap<>();

        /**
         * onCreate event.
         */
        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            addPreferencesFromResource(R.xml.pref_settings);

            // App info
            findPreference(getString(R.string.prefkey_application_details)).setOnPreferenceClickListener(applicationDetailsPreferenceClickListener);
            // About
            findPreference(getString(R.string.prefkey_about)).setOnPreferenceClickListener(aboutPreferenceClickListener);

            initSummary(getPreferenceScreen());
        }
        /**
         * onResume event.
         */
        @Override
        public void onResume() {
            super.onResume();
            getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(listener);
        }

        /**
         * onPause event.
         */
        @Override
        public void onPause() {
            super.onPause();
            getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(listener);
        }



        /**
         * App info.
         */
        private Preference.OnPreferenceClickListener applicationDetailsPreferenceClickListener = new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                Intent intent = new Intent();
                intent.setAction(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                intent.setData(Uri.parse("package:" + getActivity().getPackageName()));
                startActivity(intent);
                return true;
            }
        };

        /**
         * About.
         */
        private Preference.OnPreferenceClickListener aboutPreferenceClickListener = new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                AboutDialogFragment.newInstance().show(getActivity());
                return true;
            }
        };

        /**
         * Initialize summary.
         * @param p Target preference.
         */
        private void initSummary(Preference p) {
            if (p == null) return;

            // Get summary length
            CharSequence summary = p.getSummary();
            if (summary != null && summary.length() > 0) {
                if (summary.toString().lastIndexOf("\n") != 0) p.setSummary(summary + "\n"); // 改行追加
                summaryLengthMap.put(p, p.getSummary().length());
            } else {
                summaryLengthMap.put(p, 0);
            }

            // Update summary
            if (p instanceof PreferenceCategory) {
                PreferenceCategory pCat = (PreferenceCategory) p;
                for (int i = 0; i < pCat.getPreferenceCount(); i++) {
                    initSummary(pCat.getPreference(i));
                }
            } else if (p instanceof PreferenceScreen) {
                PreferenceScreen ps = (PreferenceScreen) p;
                for (int i = 0; i < ps.getPreferenceCount(); i++) {
                    initSummary(ps.getPreference(i));
                }
            } else {
                updatePrefSummary(p);
            }
        }

        /**
         * Update summary.
         * @param p Target preference.
         */
        private void updatePrefSummary(Preference p) {
            if (p == null) return;

            String key = p.getKey();
            CharSequence summary = p.getSummary();
            if (TextUtils.isEmpty(key)) return;
            if (TextUtils.isEmpty(summary)) summary = "";

            // preference type
            if (p instanceof ListPreference) {
                // ListPreference
                ListPreference pref = (ListPreference) p;
                pref.setValue(p.getSharedPreferences().getString(pref.getKey(), "")); // set value at once
                p.setSummary(summary.subSequence(0, summaryLengthMap.get(p)) + getString(R.string.settings_summary_current_value, pref.getEntry()));
            } else if (p instanceof MultiSelectListPreference) {
                // MultiSelectListPreference
                MultiSelectListPreference pref = (MultiSelectListPreference) p;
                Set<String> stringSet = pref.getSharedPreferences().getStringSet(pref.getKey(), null);
                String text = "";
                if (stringSet != null && stringSet.size() > 0) {
                    pref.setValues(stringSet); // set value at once
                    StringBuilder builder = new StringBuilder();
                    for (int i = 0; i < pref.getEntries().length; i++) {
                        if (stringSet.contains(pref.getEntryValues()[i])) {
                            builder.append(pref.getEntries()[i]).append(",");
                        }
                    }
                    if (builder.length() > 0) {
                        text = builder.substring(0, builder.length() - 1); // delete end comma
                    }
                }
                p.setSummary(summary.subSequence(0, summaryLengthMap.get(p)) + getString(R.string.settings_summary_current_value, text));
            } else if (p instanceof EditTextPreference) {
                // EditTextPreference
                EditTextPreference pref = (EditTextPreference) p;
                String text = p.getSharedPreferences().getString(pref.getKey(), ""); // don't use pref.getText() because of update failed.

                // adjust number
                int inputType = pref.getEditText().getInputType();
                try {
                    if ( (inputType & InputType.TYPE_CLASS_NUMBER) > 0) {
                        if ((inputType & InputType.TYPE_NUMBER_FLAG_DECIMAL) > 0) {
                            // real number
                            float val = Float.valueOf(text);
                            if ((inputType & InputType.TYPE_NUMBER_FLAG_SIGNED) == 0 && val < 0) {
                                val = 0;
                            }
                            text = String.valueOf(val);
                        } else {
                            // digit number
                            int val = Integer.valueOf(text);
                            if ((inputType & InputType.TYPE_NUMBER_FLAG_SIGNED) == 0 && val < 0) {
                                val = 0;
                            }
                            text = String.valueOf(val);
                        }
                    }
                } catch (NumberFormatException e) {
                    text = "0";
                }
                pref.setText(text); // set value at once
                p.setSummary(summary.subSequence(0, summaryLengthMap.get(p)) + getString(R.string.settings_summary_current_value, text));
            }
        }

        /**
         * On Shared Preference Change Listener.
         */
        private SharedPreferences.OnSharedPreferenceChangeListener listener = new SharedPreferences.OnSharedPreferenceChangeListener() {
            @Override
            public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
                // Update summary
                updatePrefSummary(findPreference(key));
            }
        };

    }

}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lyricsscraper/activity/SettingsActivity.kt
package com.wa2c.android.medoly.plugin.action.lyricsscraper.activity

import android.os.Bundle
import android.preference.PreferenceActivity
import android.view.MenuItem
import com.wa2c.android.medoly.plugin.action.lyricsscraper.R


/**
 * Settings activity.
 */
class SettingsActivity : PreferenceActivity() {

    /**
     * onCreate
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)

        // Action bar
        actionBar.setDisplayShowHomeEnabled(true)
        actionBar.setDisplayHomeAsUpEnabled(true)
        actionBar.setDisplayShowTitleEnabled(true)
        actionBar.setTitle(R.string.title_activity_settings)
    }

    /**
     * onOptionsItemSelected
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                finish()
                return true
            }
        }

        return super.onOptionsItemSelected(item)
    }


//    /**
//     * onCreate event.
//     */
//    override fun onCreate(savedInstanceState: Bundle?) {
//        super.onCreate(savedInstanceState)
//        fragmentManager.beginTransaction().replace(android.R.id.content, SettingsFragment()).commit()
//
//        // Action bar
//        val actionBar = actionBar
//        if (actionBar != null) {
//            actionBar.setDisplayShowHomeEnabled(true)
//            actionBar.setDisplayHomeAsUpEnabled(true)
//            actionBar.setDisplayShowTitleEnabled(true)
//            actionBar.setTitle(R.string.title_activity_settings)
//        }
//    }
//
//    /**
//     * onOptionsItemSelected event.
//     */
//    override fun onOptionsItemSelected(item: MenuItem): Boolean {
//        when (item.itemId) {
//            android.R.id.home -> {
//                finish()
//                return true
//            }
//        }
//
//        return super.onOptionsItemSelected(item)
//    }
//
//
//    /**
//     * Settings fragment.
//     */
//    class SettingsFragment : PreferenceFragment() {
//
//
//        /**
//         * App info.
//         */
//        private val applicationDetailsPreferenceClickListener = Preference.OnPreferenceClickListener {
//            val intent = Intent()
//            intent.action = Settings.ACTION_APPLICATION_DETAILS_SETTINGS
//            intent.data = Uri.parse("package:" + activity.packageName)
//            startActivity(intent)
//            true
//        }
//
//        /**
//         * About.
//         */
//        private val aboutPreferenceClickListener = Preference.OnPreferenceClickListener {
//            AboutDialogFragment.newInstance().show(activity)
//            true
//        }
//
//        /**
//         * On Shared Preference Change Listener.
//         */
//        private val listener = SharedPreferences.OnSharedPreferenceChangeListener { sharedPreferences, key ->
//            // Update summary
//            updatePrefSummary(findPreference(key))
//        }
//
//        /**
//         * onCreate event.
//         */
//        override fun onCreate(savedInstanceState: Bundle?) {
//            super.onCreate(savedInstanceState)
//            addPreferencesFromResource(R.xml.pref_settings)
//
//            // App info
//            findPreference(getString(R.string.prefkey_application_details)).onPreferenceClickListener = applicationDetailsPreferenceClickListener
//            // About
//            findPreference(getString(R.string.prefkey_about)).onPreferenceClickListener = aboutPreferenceClickListener
//
//            initSummary(preferenceScreen)
//        }
//
//        /**
//         * onResume event.
//         */
//        override fun onResume() {
//            super.onResume()
//            preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(listener)
//        }
//
//        /**
//         * onPause event.
//         */
//        override fun onPause() {
//            super.onPause()
//            preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(listener)
//        }
//
//        /**
//         * Initialize summary.
//         * @param p Target preference.
//         */
//        private fun initSummary(p: Preference?) {
//            if (p == null) return
//
//            // Get summary length
//            val summary = p.summary
//            if (summary != null && summary.length > 0) {
//                if (summary.toString().lastIndexOf("\n") != 0) p.summary = summary.toString() + "\n" // 改行追加
//                summaryLengthMap[p] = p.summary.length
//            } else {
//                summaryLengthMap[p] = 0
//            }
//
//            // Update summary
//            if (p is PreferenceCategory) {
//                val pCat = p as PreferenceCategory?
//                for (i in 0 until pCat!!.preferenceCount) {
//                    initSummary(pCat!!.getPreference(i))
//                }
//            } else if (p is PreferenceScreen) {
//                val ps = p as PreferenceScreen?
//                for (i in 0 until ps!!.preferenceCount) {
//                    initSummary(ps!!.getPreference(i))
//                }
//            } else {
//                updatePrefSummary(p)
//            }
//        }
//
//        /**
//         * Update summary.
//         * @param p Target preference.
//         */
//        private fun updatePrefSummary(p: Preference?) {
//            if (p == null) return
//
//            val key = p.key
//            var summary = p.summary
//            if (TextUtils.isEmpty(key)) return
//            if (TextUtils.isEmpty(summary)) summary = ""
//
//            // preference type
//            if (p is ListPreference) {
//                // ListPreference
//                val pref = p as ListPreference?
//                pref!!.value = p.sharedPreferences.getString(pref.key, "") // set value at once
//                p.setSummary(summary.subSequence(0, summaryLengthMap[p]).toString() + getString(R.string.settings_summary_current_value, pref.entry))
//            } else if (p is MultiSelectListPreference) {
//                // MultiSelectListPreference
//                val pref = p as MultiSelectListPreference?
//                val stringSet = pref!!.sharedPreferences.getStringSet(pref.key, null)
//                var text = ""
//                if (stringSet != null && stringSet.size > 0) {
//                    pref.values = stringSet // set value at once
//                    val builder = StringBuilder()
//                    for (i in 0 until pref.entries.size) {
//                        if (stringSet.contains(pref.entryValues[i])) {
//                            builder.append(pref.entries[i]).append(",")
//                        }
//                    }
//                    if (builder.length > 0) {
//                        text = builder.substring(0, builder.length - 1) // delete end comma
//                    }
//                }
//                p.setSummary(summary.subSequence(0, summaryLengthMap[p]).toString() + getString(R.string.settings_summary_current_value, text))
//            } else if (p is EditTextPreference) {
//                // EditTextPreference
//                val pref = p as EditTextPreference?
//                var text = p.sharedPreferences.getString(pref!!.key, "") // don't use pref.getText() because of update failed.
//
//                // adjust number
//                val inputType = pref.editText.inputType
//                try {
//                    if (inputType and InputType.TYPE_CLASS_NUMBER > 0) {
//                        if (inputType and InputType.TYPE_NUMBER_FLAG_DECIMAL > 0) {
//                            // real number
//                            var `val` = java.lang.Float.valueOf(text)!!
//                            if (inputType and InputType.TYPE_NUMBER_FLAG_SIGNED == 0 && `val` < 0) {
//                                `val` = 0f
//                            }
//                            text = `val`.toString()
//                        } else {
//                            // digit number
//                            var `val` = Integer.valueOf(text)!!
//                            if (inputType and InputType.TYPE_NUMBER_FLAG_SIGNED == 0 && `val` < 0) {
//                                `val` = 0
//                            }
//                            text = `val`.toString()
//                        }
//                    }
//                } catch (e: NumberFormatException) {
//                    text = "0"
//                }
//
//                pref.text = text // set value at once
//                p.setSummary(summary.subSequence(0, summaryLengthMap[p]).toString() + getString(R.string.settings_summary_current_value, text))
//            }
//        }
//
//        companion object {
//
//            /** Summary length.  */
//            private val summaryLengthMap = LinkedHashMap<Preference, Int>()
//        }
//
//    }

}
